python/ray/tune/schedulers/hyperband.py

```diff
@@ -250,18 +251,30 @@ def _process_bracket(
                 bracket.cleanup_full(trial_runner)
                 return TrialScheduler.STOP
 
+            bracket.is_being_processed = True
+
             good, bad = bracket.successive_halving(self._metric, self._metric_op)
+
+            logger.debug(
+                f"Processing {len(good)} good and {len(bad)} bad trials in "
+                f"bracket {bracket}.\n"
+                f"Good: {good}\nBad: {bad}"
+            )
+
             # kill bad trials
             self._num_stopped += len(bad)
             for t in bad:
                 if t.status == Trial.PAUSED:
+                    logger.debug(f"Stopping other trial {str(t)}")
                     trial_runner.stop_trial(t)
                 elif t.status == Trial.RUNNING:
+                    logger.debug(f"Stopping current trial {str(t)}")
                     bracket.cleanup_trial(t)
                     action = TrialScheduler.STOP
                 else:
                     raise TuneError(
-                        f"Trial with unexpected bad status " f"encountered: {t.status}"
+                        f"Trial with unexpected bad status encountered: "
+                        f"{str(t)} is {t.status}"
                     )
```

Comment:
Will there be a situation where a trial is marked good, left in a pending state, then gets switched to being a bad trial?

I'm thinking there is still some gap of time between the call to `trial_runner._set_trial_status(t, Trial.PENDING)` and the actual `actor_started` callback setting the status to `RUNNING`. In this time, could `process_bracket` be called again - leading to the same issue.

Reply:
In that case the bracket shouldn't get processed as the trial in question didn't report any new results, so couldn't reach the time threshold.

I actually think that the relaxation in #35338 for good trials (allowing PENDING) is not necessary anymore, but will investigate this separately